### PR TITLE
feat(apis_entities): copy fields when merging entities

### PR DESCRIPTION
**Pull request by b1rger**
This commit introduces a `merge_fields` method in the TempEntityClass.
It copies information for the relevant fields in TempEntityClass and can
be extended by subclasses to implement custom logic for copying custom
subclass fields.

Closes: #140

### CODE REVIEW

**Comment by b1rger** [comment]
Please note that I made some decisions about how attributes are merged- i.e. the `name` field gets concatenated with a comma. There might by better approaches.

**Comment by koeaw** [comment]
@b1rger I'm not entirely clear on the long(er)-term purpose of this seeing as we're about to drop `TempEntityClass`. Is this supposed to be moved elsewhere later on? Or can this help with _moving away_ from `TempEntityClass`?

I had a look at the linked issue but am none the wiser. It's possible I'm only missing a bit of context here.

**Comment by b1rger** [comment]
> @b1rger I'm not entirely clear on the long(er)-term purpose of this seeing as we're about to drop `TempEntityClass`. Is this supposed to be moved elsewhere later on? Or can this help with _moving away_ from `TempEntityClass`?
> 
> I had a look at the linked issue but am none the wiser. It's possible I'm only missing a bit of context here.

At the moment when entities are merged information gets lost and there is no way for ontologies to define how the local fields of their entities should be handled when merging is done. I don't believe that `TempEntityClass` will be going away that soon, so I think it makes sense to fix this problem

File: apis_core/apis_entities/models.py [hunk 1]
```diff
@@ -274,6 +274,26 @@ def get_duplicate_url(self):
...
         )
 
+    def merge_fields(self, other):
+        """
+        This method is intended to be extended by subclasses.
+        It is called by the `merge_with` method and should transfer data from `other` to `self`
+        """
+        # Inherited from RootObject
+        if self.name != other.name:
+            self.name = ", ".join([self.name, other.name])
```
Comment (sennierer) [inline apis_core/apis_entities/models.py:284]: I think it should be rather
```python
self.name = f"{self.name} ({other.name})"
```
for the dates `self` is overruling `other` when both objects have dates, that should be reflected in the string based fields as well

File: apis_core/apis_entities/models.py [hunk 2]
```diff
@@ -274,6 +274,26 @@ def get_duplicate_url(self):
...
+        # Boolean values
+        self.review = self.review and other.review
+        self.status = self.status and other.status
+        self.published = self.published and other.published
+        # date strings: we use the other date only if the self date is missing
+        self.start_date_written = self.start_date_written or other.start_date_written
+        self.end_date_written = self.end_date_written or other.end_date_written
+        # we simply concatenate strings
+        self.references = "\n".join(filter(None, [self.references, other.references]))
```
Comment (sennierer) [inline apis_core/apis_entities/models.py:293]: see comment above. should be clear which comments stem from an object that was merged in the actual one (put in brackets or explicitly mentioned ... eg `references from merged objects:`

**Review by sennierer — CHANGES_REQUESTED** [review]
I think the merging is currently not completely concise:
`self` is overruling `other` on some fields, others are completely equal.
I think the object that something is merged into (so `self`) should be overruling `other` and this should also be reflected in the string based fields.

Given that this should work on all classes, so include also fields that are not inheritated from `RootObject` we could refactor that to run through all fields and depending on the field type (so whether its sting, date etc. always use the same logic/function on it.

**Comment by b1rger** [comment]
> Given that this should work on all classes, so include also fields that are not inheritated from RootObject we could refactor that to run through all fields and depending on the field type (so whether its sting, date etc. always use the same logic/function on it.

Oke, I've implemented that - the `merge_fields` method now iterates through the model fields and checks wether there is a `merge_{fieldname}` method and if so uses that. If not, it checks for a `merge_{fieldtype}` method and uses that. If no method exists and the field in `self` is empty, it copies the value of `other`.

I've implemented `merge_` methods for charfield, textfield and booleanfield. Given that `start_date_written` and `end_date_written` are parsed and the parsing would fail with the default merge method for charfields, I added extra methods for those two, demonstrating how the merge methods get chosen.

In the long run, some of the merge functionality should probably move to the RootObject or the AbstractEntity...

**Pull request merged**